server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1708,8 +1739,26 @@ synchronized void addShardDeleteResult(
         ) {
             try {
                 shardGenerationsBuilder.put(indexId, shardId, newGeneration);
-                new ShardSnapshotMetaDeleteResult(Objects.requireNonNull(indexId.getId()), shardId, blobsToDelete).writeTo(compressed);
-                resultCount += 1;
+                // There is a minimum of 1 byte available for writing
+                if (this.truncatedShardDeleteResultsOutputStream.hasCapacity()) {
+                    new ShardSnapshotMetaDeleteResult(Objects.requireNonNull(indexId.getId()), shardId, blobsToDelete).writeTo(compressed);
+                    // We only want to read this shard delete result if we were able to write the entire object.
+                    // Otherwise, for partial writes, an EOFException will be thrown upon reading
+                    if (this.truncatedShardDeleteResultsOutputStream.hasCapacity()) {
+                        successfullyWrittenBlobsCount += 1;
+                    } else {
+                        leakedBlobsCount += 1;
```

Comment:
Likewise this is recording the number of shards with leaked blobs rather than the number of leaked blobs. However, rather than just renaming the variable I think we should actually count the number of leaked blobs (i.e. `+= blobsToDelete.size()` here).